### PR TITLE
D8CORE-4339 Allow restriction based on the person's affiliation to the university

### PR DESCRIPTION
**Pull request by pookmish**
# READY FOR REVIEW

# Summary
- Add configurable restriction based on the affiliation.

# Need Review By (Date)
- 6/3

# Urgency
- medium

# Steps to Test
1. Follow steps in https://github.com/SU-SWS/stanford_profile/pull/424

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

File: config/schema/stanford_ssp.schema.yml [hunk 1]
```diff
@@ -24,15 +24,25 @@ stanford_ssp.settings:
...
-      sequence:
-        type: string
-        label: 'Allowed User'
+    allowed:
+      type: mapping
+      label: 'Allowed Authorizations'
+      mapping:
+        affiliations:
+          type: sequence
+          label: 'Allowed Groups'
```
Comment (sherakama) [inline config/schema/stanford_ssp.schema.yml:33]:
```suggestion
          label: 'Allowed Affiliations'
```

File: config/schema/stanford_ssp.schema.yml [hunk 2]
```diff
@@ -24,15 +24,25 @@ stanford_ssp.settings:
...
+          sequence:
+            type: string
+            label: 'Allowed Group'
```
Comment (sherakama) [inline config/schema/stanford_ssp.schema.yml:36]:
```suggestion
            label: 'Allowed Affiliations'
```

File: stanford_ssp.module [hunk 3]
```diff
@@ -96,18 +96,28 @@ function stanford_ssp_simplesamlphp_auth_allow_login($attributes) {
...
   $authname = $auth_manager->getAuthname();
 
-  if (in_array($authname, $config->get('allowed_users'))) {
+  // Check if the user name is allowed.
+  if (in_array($authname, $config->get('allowed.users'))) {
     return TRUE;
   }
 
+  // Check if the user is an allowed affiliation.
+  if (!empty($attributes['eduPersonAffiliation']) && array_intersect($config->get('allowed.affiliations'), $attributes['eduPersonAffiliation'])) {
```
Comment (sherakama) [inline stanford_ssp.module:105]: Any reason we might want `eduPersonAffiliation` to be a configurable SAML attribute here?

File: tests/src/Kernel/Form/AuthorizationsFormTest.php [hunk 4]
```diff
@@ -30,15 +30,15 @@ class AuthorizationsFormTest extends KernelTestBase {
   protected function setUp(): void {
     parent::setup();
     \Drupal::configFactory()->getEditable('stanford_ssp.settings')
-      ->set('allowed_groups', [])
-      ->set('allowed_users', [])
+      ->set('allowed.groups', [])
```
Comment (sherakama) [inline tests/src/Kernel/Form/AuthorizationsFormTest.php:33]: Add affiliations

File: tests/src/Kernel/Form/AuthorizationsFormTest.php [hunk 5]
```diff
@@ -57,10 +57,10 @@ public function testForm() {
     \Drupal::formBuilder()
       ->submitForm('\Drupal\stanford_ssp\Form\AuthorizationsForm', $form_state);
 
-    $this->assertTrue(in_array('group1', \Drupal::config('stanford_ssp.settings')->get('allowed_groups')));
-    $this->assertTrue(in_array('group2', \Drupal::config('stanford_ssp.settings')->get('allowed_groups')));
-    $this->assertTrue(in_array('user1', \Drupal::config('stanford_ssp.settings')->get('allowed_users')));
-    $this->assertTrue(in_array('user2', \Drupal::config('stanford_ssp.settings')->get('allowed_users')));
+    $this->assertTrue(in_array('group1', \Drupal::config('stanford_ssp.settings')->get('allowed.groups')));
```
Comment (sherakama) [inline tests/src/Kernel/Form/AuthorizationsFormTest.php:60]: Add affiliations test

File: stanford_ssp.module (hunk 3 above)
Comment (pookmish) [reply stanford_ssp.module:105]: I don't think we need to at this time.

File: src/Form/AuthorizationsForm.php [hunk 6]
```diff
@@ -52,19 +54,38 @@ public function buildForm(array $form, FormStateInterface $form_state) {
       ],
     ];
 
+    $url = Url::fromUri('https://uit.stanford.edu/service/saml/arp/edupa');
+    $affiliation_link = Link::fromTextAndUrl('SAML Affiliation Information', $url)
+      ->toString();
+    $form['allowed_affiliations'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Affiliation'),
+      '#description' => $this->t("Restrict to the user's affiliation to the Stanford. View what these affiliations entail at @link.", ['@link' => $affiliation_link]),
```
Comment (imonroe) [inline src/Form/AuthorizationsForm.php:63]:
```suggestion
      '#description' => $this->t("Restrict to the user's affiliation to Stanford. View what these affiliations entail at @link.", ['@link' => $affiliation_link]),
```

**Review by imonroe — APPROVED** [review]
One small change, then I'm good with it.  Tested with https://github.com/SU-SWS/stanford_profile/pull/424/files and it functions correctly.

**Pull request merged**